enhancements/microshift/loadbalancer-service-support.md

```diff
@@ -0,0 +1,199 @@
+differently between OVN-Kubernetes and other CNIs.
+
+## Design Details
+
+The Service of loadblancer type is designed to exposes the Service externally
+using a cloud provider's load balancer. Normally, there is a loadbalancer
+Service controller which is responsible for:
+
+1. provision a loadbalancer for the Service.
+2. publish the ingress endpoint information in the Service's
+   `.status.loadBalancer` field.
+
+However, in the context of Microshift, as we only get one node in the cluster,
+it doesn't make sense to use a real load balancer to forward the ingress
+traffic. So instead of provisioning the loadbalancer instance, the controller
+is only responsible for helping the CNIs to plumbing the ingress traffic.
+
+### Traffic Plumbing
+
+#### OVN-Kubernetes
+
+OVN-Kubernetes has already implemented the [loadbalancer ingress
+IP](https://github.com/openshift/ovn-kubernetes/blob/master/docs/external-ip-and-loadbalancer-ingress.md)
+support. The ovnkube-node component which is running on each host watches the
+Service's `.status.loadBalancer` field, and insert iptables rules that does DNAT
+to the clusterIP of the Service. Therefore, from the loadbalancer Service
+controller perspective, it will only be responsible for the updating the status
+of the Service. The ovn-kubernetes will do the rest of the job.
+
+#### Other CNIs
+
+For other CNIs, which don't have the similar function as OVN-Kubernetes, neither
+do they bypass the kernel for the ingress traffic, we can take the similar
+approach as K3S. The K3S's loadbalancer Service controller provisions a
+[klipper-lb](https://github.com/k3s-io/klipper-lb) pod to each nodes for each
+loadbalancer Service. The pod can insert iptables rules which DNAT ingress
+traffic to the nodeIP of the Service.
+
+
+### Update Status
+As, we don't want to ask user to allocated additional VIP as the loadbalancer
+IP. We can say that the ingress ip of the Service's `.status.loadBalancer` field
+shall always be the existing IPs of the node. When there are multiple IPs, the
+kubelet `NodeIP` shall be used by default.
+
+The controller shall also check if the Service port has already been allocated
+to other loadbalancer Service before updating the Service status.
+
+### Open Questions [optional]
+
+* Do we want to allow users to choose the IP other than the NodeIP as the
+Ingress IP of loadbalancer Service, when there are multiple IPs available?
+
+### Test Plan
+
+N/A
+
+### Graduation Criteria
+
+N/A
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+N/A
+
+### Operational Aspects of API Extensions
+
+N/A
+
+#### Failure Modes
+
+If the port of one Service  has already been occupied by either a node process
+or other loadbalancer Services, the controller shall leave the
+`.status.loadBalancer` of the Service blank, thus the Service will be put in
```

Comment:
Is it possible to set it in error state?, I think it's better if we can explain the user that the port is busy somewhere in the system and that's why it's failing.

Reply:
In k8s cloud provider implementation, it will create an event when an error happens. I think we take the same approach.
https://github.com/kubernetes/cloud-provider/blob/73b07b952335721870a60aebfc8936af19fe921b/controllers/service/controller.go#L327